### PR TITLE
cmd/{containerboot,k8s-operator}: store proxy device ID early to help with cleanup for broken proxies

### PR DESCRIPTION
**Pull request by irbekrm**
## Background

For Kubernetes operator proxies, the proxy device ID, tailnet IPs and MagicDNS name are stored in a Kubernetes `Secret`.
Device ID is used [when cleaning up a device](https://github.com/tailscale/tailscale/blob/main/cmd/k8s-operator/sts.go#L228-L234) and tailnet IPs and MagicDNS name are stored on resources that users interact with for visibility (i.e [`Ingress`'s status](https://github.com/tailscale/tailscale/blob/main/cmd/k8s-operator/ingress.go#L278-L295)).
At the moment, all three are stored at the same time, once the proxy route setup has succeeded.
This means that if proxy route setup fails, the operator won't be able to clean up the device, see #12146 .

## Changes in this PR

This PR splits the device info storing - we now store device ID early to ensure that it's availabe even if route setup has failed.

## Notes

I have tested that for a proxy where the route setup cannot succeed, the device ID (but not IPs/MagicDNS name) get stored and that device cleanup succeeds:

<details>
<code>
// 1. Create a  tailnet egress Service with a non-existant FQDN:
$ kubectl get svc testcleanup -oyaml
apiVersion: v1
kind: Service
metadata:
  annotations:
    tailscale.com/hostname: testcleanup
    tailscale.com/tailnet-fqdn: xyz.tailfake.ts.net
  name: testcleanup
spec:
  externalName: placeholder
  type: ExternalName

// 2. Observe that the state Secret for this has `device_id`, but not `device_fqdn` or `device_ips` fields:
$ kubectl view-secret ts-testcleanup-jtvvv-0 -n tailscale
-> _profiles
-> cap-95.hujson
-> device_id
-> profile-6430
-> tailscaled
-> _current-profile
-> _machinekey
// 3. Delete the Service and observe (in the Admin panel) that the associated Tailscale device has been deleted
</code>
</details>

Updates tailscale/tailscale#12146

### CODE REVIEW

File: cmd/containerboot/main.go [hunk 1]
```diff
@@ -321,7 +321,7 @@ authLoop:
 		}
 	}
 
-	if cfg.InKubernetes && cfg.KubeSecret != "" && cfg.KubernetesCanPatch && isTwoStepConfigAuthOnce(cfg) {
+	if hasKubeStateStore(cfg) && isTwoStepConfigAuthOnce(cfg) {
```
Comment (irbekrm) [inline cmd/containerboot/main.go:324]: small drive-by cleanup, not a functional change

File: cmd/containerboot/main.go [hunk 2]
```diff
@@ -337,11 +337,10 @@ authLoop:
 	}
 
 	var (
-		wantProxy         = cfg.ProxyTargetIP != "" || cfg.ProxyTargetDNSName != "" || cfg.TailnetTargetIP != "" || cfg.TailnetTargetFQDN != "" || cfg.AllowProxyingClusterTrafficViaIngress
-		wantDeviceInfo    = cfg.InKubernetes && cfg.KubeSecret != "" && cfg.KubernetesCanPatch
```
Comment (irbekrm) [inline cmd/containerboot/main.go:341]: small drive-by cleanup, not a functional change

File: cmd/containerboot/main.go [hunk 3]
```diff
@@ -355,7 +354,7 @@ authLoop:
 		go watchServeConfigChanges(ctx, cfg.ServeConfigPath, certDomainChanged, certDomain, client)
 	}
 	var nfr linuxfw.NetfilterRunner
-	if wantProxy {
+	if isTCPProxy(cfg) {
```
Comment (irbekrm) [inline cmd/containerboot/main.go:357]: small drive-by cleanup, not a functional change

File: cmd/k8s-operator/sts.go [hunk 4]
```diff
@@ -424,7 +426,12 @@ func (a *tailscaleSTSReconciler) DeviceInfo(ctx context.Context, childLabels map
 	// to remove it.
 	hostname = strings.TrimSuffix(string(sec.Data["device_fqdn"]), ".")
 	if hostname == "" {
-		return "", "", nil, nil
+		// Device ID gets stored and retrieved in a different flow than
+		// FQDN and IPs. A device that acts as Kubernetes operator
+		// proxy, but whose route setup has failed might have an device
+		// ID, but no FQDN/IPs. If so, return the ID, to allow the
+		// operator to clean up such devices.
+		return id, "", nil, nil
```
Comment (irbekrm) [inline cmd/k8s-operator/sts.go:434]: This is fine, the id is only read by clean up functionality who does actually need it

File: cmd/containerboot/main.go [hunk 5]
```diff
@@ -1287,6 +1321,19 @@ func isOneStepConfig(cfg *settings) bool {
 	return cfg.TailscaledConfigFilePath != ""
 }
 
+// isTCPProxy returns true if the Tailscale node needs to be configured to act
+// as a TCP proxy, proxying to an endpoint provided via one of the config env
+// vars.
+func isTCPProxy(cfg *settings) bool {
```
Comment (tomhjp) [inline cmd/containerboot/main.go:1327]: Maybe `isL3Proxy` to save confusion about UDP?

File: cmd/containerboot/kube.go [hunk 6]
```diff
@@ -19,22 +19,44 @@ import (
...
 			if s.Code >= 400 && s.Code <= 499 {
-				// Assume the secret doesn't exist, or we don't have
-				// permission to access it.
+				// Assume the Secret doesn't exist, or we don't
+				// have permissions to access it.
 				return nil
 			}
 		}
 		return err
 	}
```
Comment (tomhjp) [inline cmd/containerboot/kube.go:36]: Maybe worth pulling this shared code out into a function like `func secretAccessible(ctx context.Context, secretName string) (bool, error)`?

**Review by tomhjp — APPROVED** [review]
LGTM, just a couple of optional suggestions

File: cmd/containerboot/main.go (hunk 5 above)
Comment (irbekrm) [reply cmd/containerboot/main.go:1327]: Yeah that would be more accurate thanks 👍🏼
(I think I started using 'TCP proxy' wording in places because we had been thinking about replacing the iptables/nftables rules with a proxy implemented on top of tailscale serve and there already is a [`tcpHandlerForServe`](https://github.com/tailscale/tailscale/blob/v1.69.0-pre/ipn/ipnlocal/serve.go#L420) that does some of what we need)

File: cmd/containerboot/kube.go (hunk 6 above)
Comment (irbekrm) [reply cmd/containerboot/kube.go:36]: Thinking about it, looks like this check is now redundant. We check that we have the right permissions to get/patch/create Secret [here](https://github.com/tailscale/tailscale/blob/v1.69.0-pre/cmd/containerboot/main.go#L192). I am going to remove this bit altogether.

Comment (tomhjp) [reply cmd/containerboot/kube.go]: Nice! 🎉

Comment (irbekrm) [reply cmd/containerboot/kube.go]: (did some additional testing to ensure that in cases, where the Secret is not precreated (and is expected to be created by kube state store used by tailscaled), this does not race with Secret creation)
